### ISSUE TITLE
Add Retry Behaviour to AWS Kinesis Data Firehose Sink

### ISSUE THREAD

**Issue by mattaltberg** [vectordotdev/vector#12835]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Currently, if any records are throttled by AWS, Vector does not retry sending the records to Firehose. We should add a retry to ensure records are sent, because right now, the data is simply dropped.

**Comment by mattaltberg** [comment]
Hey @jszwedko, I'm using 0.16.1. I'll ping you again after I try using the latest version.

**Comment by jszwedko** [comment]
Thanks @mattaltberg . I am curious about the newest version (0.21.2) since there were some changes to switch to the new AWS SDK and handle throttling responses consistently in 0.21.0.

**Comment by mattaltberg** [comment]
@jszwedko I tested using 0.21.2, and still got the same throttling issue with my Firehose stream. Is it possible it's not used for the Firehose sink?

**Comment by jszwedko** [comment]
If you are able to re-run with the environment variable `VECTOR_LOG=aws_smithy_http=trace` that should output trace messages from the AWS SDK that show the responses coming from AWS.

**Comment by jszwedko** [comment]
Interesting, thanks for sharing the response @mattaltberg ! It seems like maybe we should be looking for `ServiceUnavailableException` in that code block I linked, too. cc/ @fuchsnj for thoughts.

**Comment by jszwedko** [comment]
Also noticing that it seems to return one response per record, in your log some of the records went in. It seems like we'll need to handle retrying only subsets of the records (similar to Elasticsearch). At the least, for an initial implementation, we could retry the full request though.

**Comment by mattaltberg** [comment]
Yeah, if you want more information, my setup is using a `unix_stream` socket, reading from `/var/log/output.sock`. It then gets transformed (new line gets added to the end of the message) before the Firehose sinks gets it.

**Comment by mattaltberg** [comment]
Hey @jszwedko any updates? I'm curious if I'll need to keep upping my Firehose quota

**Comment by dengmingtong** [comment]
Is this issue already fixed? Or is there any plan to fix it?

**Comment by jszwedko** [comment]
It's not fixed yet, but a contributor has a PR open for it: https://github.com/vectordotdev/vector/pull/16771
